stasis_broadcast: Add optional ARI broadcast with first-claim-wins

danieldonoghue opened the pull request:
Adds two optional modules:
res_stasis_broadcast.so: Infrastructure for broadcasting a single incoming
channel to multiple ARI applications with atomic first-claim-wins semantics.

app_stasis_broadcast.so: Provides the StasisBroadcast() dialplan application
which invokes the broadcast infrastructure.

Both modules are self-contained; if neither is loaded there is zero runtime
impact. Loading them does not alter existing Stasis or ARI behavior unless
explicitly used.

Key Features (only active when modules are loaded):
Fisher-Yates shuffled broadcast dispatch for fair claim races
Atomic claim operations using mutex + condition variable signaling
Configurable broadcast timeouts
Safe regex application filtering with validation to mitigate ReDoS risk
Thread-safe channel variable snapshotting (channel locked during reads)
Late-claim safety: broadcast context kept alive until after the Stasis
session ends so concurrent claimants always receive 409 Conflict rather
than 404 Not Found
Memory safety via RAII_VAR, ast_json_ref/unref, and ao2 reference counting

Components Added:
res/res_stasis_broadcast.c: Core broadcast + claim logic
apps/app_stasis_broadcast.c: StasisBroadcast() dialplan application
include/asterisk/stasis_app_broadcast.h: Public API header
res/ari/resource_events.c: Integrates POST /ari/events/claim endpoint
rest-api/api-docs/events.json: New CallBroadcast and CallClaimed events

Implementation Notes:
Broadcast contexts reside in an ao2 hash container keyed by channel id. Each
context holds atomic claim state, winner application name, timeout metadata,
and a condition variable for waiters. Broadcast contexts are kept alive until
after stasis_app_exec() returns so that concurrent claimants racing against
the timeout always receive 409 Conflict. Broadcast dispatch calls
stasis_app_send() directly for each matching application in shuffled order.
Regex filters are validated with bounded length, group depth, quantified
group count, and alternation limits to reduce pathological backtracking.
Timeout calculation uses timespec arithmetic with overflow-safe millisecond
remainder handling. Event JSON follows existing Stasis/ARI conventions;
references are managed correctly to avoid leaks or double frees.

Optional Nature / Impact:
No changes to existing APIs, events, or applications when absent.
Clean fallback: systems ignoring the modules behave identically to prior
versions.

Development was assisted by Claude (Anthropic). All generated code has been
reviewed, tested, and is understood by the author.

UserNote: New optional modules res_stasis_broadcast.so and
app_stasis_broadcast.so enable broadcasting an incoming channel to multiple
ARI applications. The first application to successfully claim (via
POST /ari/events/claim) wins channel control. StasisBroadcast() dialplan
application initiates broadcasts. CallBroadcast and CallClaimed events notify
applications. When modules are not loaded, behavior is unchanged.

DeveloperNote: New public APIs in stasis_app_broadcast.h:
stasis_app_broadcast_channel(), stasis_app_claim_channel(),
stasis_app_broadcast_winner(), and stasis_app_broadcast_wait(). New ARI event
types (CallBroadcast, CallClaimed) added to events.json. All code is isolated;
no existing ABI modified.

danieldonoghue commented:
> Can you explain the reasoning behind an approach of having a separate dialplan application that does the broadcast and then returns to the dialplan, instead of a single dialplan application that makes the operation self contained?

Hi @jcolp, my reasoning was fairly simple. I wanted to create a capability that would enhance the existing Stasis application without interfering with its existing functionality and making maintenance more difficult. It was easier to reason about the development work with a separate dialplan application (and how it would work in an actual dialplan) and I could leave Stasis doing what Stasis already does very well. 

It also felt as though it might be useful to be able to do additional handling in the dialplan following the broadcast - for example, handling timeouts by relaxing app_filter or trying a fallback of some sort.

I could have modified the Stasis application but found it difficult to reason about integrating the optional arguments with it and how to explain the functionality and increased complexity, in addition to the reasons I just outlined.

I suppose I could also have embedded the Stasis application call directly into this application but worried that would make it more difficult to maintain, moving forward, should the Stasis application's parameters change and still more difficult to reason about how that would look in the dialplan and the arguments being passed through. This option became less appealing the more I thought about it.

If you feel that this is something that should become part of the standard Stasis application I'd be happy to take another look and try to integrate it directly.

jcolp commented on apps/app_stasis_broadcast.c:
```diff
@@ -0,0 +1,217 @@
+/*
+ * Asterisk -- An open source telephony toolkit.
+ *
+ * Copyright (C) 2025, Aurora Innovation AB
```
2026

jcolp commented on apps/app_stasis_broadcast.c:
```diff
@@ -0,0 +1,217 @@
+		/* Channel was claimed */
+		winner = stasis_app_broadcast_winner(ast_channel_uniqueid(chan));
+		if (winner) {
+			ast_log(LOG_NOTICE, "Channel %s claimed by %s\n",
```
debug message

jcolp commented on include/asterisk/stasis_app_broadcast.h:
```diff
@@ -0,0 +1,112 @@
+/*
+ * Asterisk -- An open source telephony toolkit.
+ *
+ * Copyright (C) 2025, Aurora Innovation AB
```
2026

jcolp commented on res/res_stasis_broadcast.c:
```diff
@@ -0,0 +1,945 @@
+/*
+ * Asterisk -- An open source telephony toolkit.
+ *
+ * Copyright (C) 2025, Aurora Innovation AB
```
2026

jcolp commented on res/res_stasis_broadcast.c:
```diff
@@ -0,0 +1,945 @@
+	/*! The unique ID of the channel */
+	char channel_id[AST_MAX_EXTENSION];
+	/*! Name of the winning application */
+	char winner_app[AST_MAX_EXTENSION];
```
I don't think there is a max for a Stasis application, and AST_MAX_EXTENSION is certainly not correct for it

jcolp commented on res/res_stasis_broadcast.c:
The unlink can be done as part of ao2_find by passing OBJ_UNLINK

danieldonoghue committed:
stasis_broadcast: Add optional ARI broadcast with first-claim-wins

Adds two optional modules:
res_stasis_broadcast.so: Infrastructure for broadcasting a single incoming
channel to multiple ARI applications with atomic first-claim-wins semantics.

app_stasis_broadcast.so: Provides the StasisBroadcast() dialplan application
which invokes the broadcast infrastructure.

Both modules are self-contained; if neither is loaded there is zero runtime
impact. Loading them does not alter existing Stasis or ARI behavior unless
explicitly used.

Key Features (only active when modules are loaded):
Fisher-Yates shuffled broadcast dispatch for fair claim races
Atomic claim operations using mutex + condition variable signaling
Configurable broadcast timeouts
Safe regex application filtering with validation to mitigate ReDoS risk
Thread-safe channel variable snapshotting (channel locked during reads)
Late-claim safety: broadcast context kept alive until after the Stasis
session ends so concurrent claimants always receive 409 Conflict rather
than 404 Not Found
Memory safety via RAII_VAR, ast_json_ref/unref, and ao2 reference counting

Components Added:
res/res_stasis_broadcast.c: Core broadcast + claim logic
apps/app_stasis_broadcast.c: StasisBroadcast() dialplan application
include/asterisk/stasis_app_broadcast.h: Public API header
res/ari/resource_events.c: Integrates POST /ari/events/claim endpoint
rest-api/api-docs/events.json: New CallBroadcast and CallClaimed events

Implementation Notes:
Broadcast contexts reside in an ao2 hash container keyed by channel id. Each
context holds atomic claim state, winner application name, timeout metadata,
and a condition variable for waiters. Broadcast contexts are kept alive until
after stasis_app_exec() returns so that concurrent claimants racing against
the timeout always receive 409 Conflict. Broadcast dispatch calls
stasis_app_send() directly for each matching application in shuffled order.
Regex filters are validated with bounded length, group depth, quantified
group count, and alternation limits to reduce pathological backtracking.
Timeout calculation uses timespec arithmetic with overflow-safe millisecond
remainder handling. Event JSON follows existing Stasis/ARI conventions;
references are managed correctly to avoid leaks or double frees.

Optional Nature / Impact:
No changes to existing APIs, events, or applications when absent.
Clean fallback: systems ignoring the modules behave identically to prior
versions.

Development was assisted by Claude (Anthropic). All generated code has been
reviewed, tested, and is understood by the author.

UserNote: New optional modules res_stasis_broadcast.so and
app_stasis_broadcast.so enable broadcasting an incoming channel to multiple
ARI applications. The first application to successfully claim (via
POST /ari/events/claim) wins channel control. StasisBroadcast() dialplan
application initiates broadcasts. CallBroadcast and CallClaimed events notify
applications. When modules are not loaded, behavior is unchanged.

DeveloperNote: New public APIs in stasis_app_broadcast.h:
stasis_app_broadcast_channel(), stasis_app_claim_channel(),
stasis_app_broadcast_winner(), and stasis_app_broadcast_wait(). New ARI event
types (CallBroadcast, CallClaimed) added to events.json. All code is isolated;
no existing ABI modified.

danieldonoghue commented:
I've removed the taskpool entirely, following the concerned tone I note from the comments. stasis_app_send() writes directly to each app's WebSocket socket synchronously anyway so the taskpool was adding overhead without any significant parallelism benefit. The application name shuffle before the dispatch loop ensures no single app is consistently first to receive the event, preserving fairness across broadcasts.

This also means the module doesnt spawn any threads, addressing the concern about idle threads for anybody loading this module.

One potential issue with removing the task pool, though, is that with the taskpool, a slow ARI client blocking on socket write would only have affected its own task thread. With this modification to serially dispatch, a client whose socket send buffer is full could delay delivery to subsequent apps in the loop. In practice this is unlikely given typical ARI client characteristics and TCP buffer sizes, but wanted to flag it.

asteriskteam commented:
This PR has been closed because it has been in "Changes Requested" or "submitter-action-required" state for more than 42 days.

danieldonoghue commented:
hmm, I see there are still failing tests. I will take a look at those as soon as possible :)

gtjoseph commented:
> hmm, I see there are still failing tests. I will take a look at those as soon as possible :)

I don't think they're related to your change.  I'll re-run them to make sure.

gtjoseph commented:
@danieldonoghue Don't worry about the two pjsip tests right now but the rest_api/authentication_user_acl test isn't one I've seen failing before.  Can you run it locally both with and without this PR and see what happens?

jcolp commented:
Is this review based on an old version of master before the ACL support went in for ARI and HTTP?

gtjoseph commented:
> Is this review based on an old version of master before the ACL support went in for ARI and HTTP?

Oh yeah, no kidding.

@danieldonoghue If you rebase this PR on current master you should be good.

danieldonoghue commented:
@gtjoseph, I did the rebase and it fixed those tests, as expected. however, I note that caller_attended fails now. looking at it, I dont believe that to be related to my changes.

gtjoseph commented:
> @gtjoseph, I did the rebase and it fixed those tests, as expected. however, I note that caller_attended fails now. looking at it, I dont believe that to be related to my changes.

Any time you see a transfer test failing there's a 99% chance it's because the test is timing dependent and GitHub is being stingy with resources. :)
